### PR TITLE
Adds a warning regarding the use of lambdas instead @Suspendable func…

### PR DESCRIPTION
**Pull request by joeldudleyr3**
…tions.

### CODE REVIEW

File: docs/source/flow-state-machines.rst [hunk 1]
```diff
@@ -188,6 +188,10 @@ and try again.
 
 .. note:: Java 9 is likely to remove this pre-marking requirement completely.
 
+.. note:: Be careful when using a lambda inside an @Suspendable function. It may capture a reference to a 
+non-suspendable Service in the ServiceHub, causing serialisation to fail. If in doubt, place the lambda in a 
+separate, non-@Suspendable function, which you can then call from the @Suspendable function.
+
```
Comment (adagys) [inline docs/source/flow-state-machines.rst]: Actually, I don't think it's got much to do with lambdas. It's probably better to say something like "Accessing the vault inside an \@Suspendable function, e.g. `serviceHub.vaultService`, might lead to serialisation errors once the fiber suspends. Place such logic in a helper non-@Suspendable function instead. We are working to fix this"

File: docs/source/flow-state-machines.rst [hunk 2]
```diff
@@ -505,4 +507,4 @@ the features we have planned:
 * Being able to interact with people, either via some sort of external ticketing system, or email, or a custom UI.
   For example to implement human transaction authorisations.
 * A standard library of flows that can be easily sub-classed by local developers in order to integrate internal
-  reporting logic, or anything else that might be required as part of a communications lifecycle.
\ No newline at end of file
+  reporting logic, or anything else that might be required as part of a communications lifecycle.
```
Comment (adagys) [inline docs/source/flow-state-machines.rst]: did anything change on this line?

Comment (joeldudleyr3) [inline docs/source/flow-state-machines.rst]: No - can't understand why its flagging up.

**Comment by mikehearn** [comment]
Services are meant to be tagged as `SerializableAsToken` or whatever it's called, to avoid this exact problem. If we have services that we forgot to tag, then we can just fix that and take out the warning. Which service was causing you problems joel?

**Comment by joeldudleyr3** [comment]
Sorry about that. It was a user who wrote the following code - it worked fine after moving the call `serviceHub.vaultService.currentVault.statesOfType<TTxState>()` to its own, non-suspendable function.

`@Suspendable`
`private fun inputListFromLastOutput(): List<StateAndRef<TTxState>> {`
`        var lastStates = serviceHub.vaultService.currentVault.statesOfType<TTxState>()`
`        if (lastStates.size == 0) {`
`            // TODO do we need notify?`
`            val genesis = initializeGenesis()`
`            serviceHub.vaultService.notify(genesis.tx)`
`            lastStates = serviceHub.vaultService.currentVault.statesOfType<TTxState>()`
`        }`
`        require(lastStates.size == 1) { "TODO GENESIS Internal error: more than one last state (last output) found." }`
`        return listOf<StateAndRef<TTxState>>(lastStates.first())`
`    }`

**Comment by mikehearn** [comment]
How did it fail? Do you have a stack trace?

**Comment by joeldudleyr3** [comment]
E 00:24:47 28 [045e1633-3c94-4d59-8cc3-e388b3252d53].uncaughtException - Caught exception from flow
     java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Was expecting a SessionInitResponse but got nothing: SendAndReceive(session=FlowSession(flow=net.corda.flows.NotaryFlow$Client@6720d3e7, otherParty=Mock Company 1, ourSessionId=6645358286741261650, otherPartySessionId=null, waitingForResponse=true), message=SessionInit(initiatorSessionId=6645358286741261650, initiatorParty=Mock Company 3, flowName=net.corda.flows.NotaryFlow$Client, firstPayload=SignRequest(tx=SignedTransaction(txBits=[120101006A6176612E7574696C2E41727261794C6973F40001000001000101016E65742E636F7264612E636F72652E636F6E7472616374732E5472616E73616374696F6E53746174E50102636F6D2E6469676974616C61737365742E706C6174666F726D2E47736C547853746174E50103636F6D2E6469676974616C61737365742E706C6174666F726D2E47736C436F6E74726163F401046E65742E636F7264612E636F72652E63727970746F2E53656375726548617368245348413235B62190FC9EA356985D4E82B4C7332537CC032B5E244C6C4CCC4DF8002C08AC9496D201056E65742E636F7264612E636F72652E63727970746F2E436F6D706F736974654B6579244C6561E60E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE07401066A6176612E7574696C2E436F6C6C656374696F6E732453696E676C65746F6E4C6973F401050E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE0740101636F7264612E6E6F746172792E73696D706C657C4D6F636B20436F6D70616E7920B101050E20CCFFF852FC48EFCC4508341574BB62D55BDE5B7538025BFB0E6F08DAE1B8D30701000101076E65742E636F7264612E636F72652E636F6E7472616374732E436F6D6D616EE40C0101086E65742E636F7264612E636F72652E63727970746F2E436F6D706F736974654B65F901050E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE0740109636F6D2E6469676974616C61737365742E706C6174666F726D2E47736C47656E65736973436F6D6D616EE414636F7264612E6E6F746172792E73696D706C657C4D6F636B20436F6D70616E7920B101050E20CCFFF852FC48EFCC4508341574BB62D55BDE5B7538025BFB0E6F08DAE1B8D307010601050E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE074010A6E65742E636F7264612E636F72652E636F6E7472616374732E5472616E73616374696F6E547970652447656E657261EC00], sigs=[[86595817C69ABFB5070F9C4FAA3B52424F08148AE39A37A8BB08837C4AE6E4870029BFF208CDC08C09A002FCA9910E6A11F6E90A33874DC965B59052161D400C], [4083C5BE0EC4BD7B16EAED7022F74669A335EECBA5AED9636B02CCB3203AE5E08A2FB59146D82D77A4722AF5669A3DC78576F2CFB34B234664944C9205D20407]], id=6CA0AA5854A10E8F341762C62290D456CD219F0D6BB9A6F1A545F76869A7C80A), callerIdentity=Mock Company 3)), receiveType=interface net.corda.node.services.statemachine.StateMachineManager$SessionInitResponse)
    	at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:91) ~[node-0.7-SNAPSHOT.jar:?]
    	at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1067) [quasar-core-0.7.6-jdk8.jar:0.7.6]
    	at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:767) [quasar-core-0.7.6-jdk8.jar:0.7.6]
    	at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100) [quasar-core-0.7.6-jdk8.jar:0.7.6]
    	at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91) [quasar-core-0.7.6-jdk8.jar:0.7.6]
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_92]
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_92]
    	at net.corda.node.utilities.AffinityExecutor$ServiceAffinityExecutor$1$thread$1.run(AffinityExecutor.kt:67) [node-0.7-SNAPSHOT.jar:?]
    Caused by: java.lang.IllegalStateException: Was expecting a SessionInitResponse but got nothing: SendAndReceive(session=FlowSession(flow=net.corda.flows.NotaryFlow$Client@6720d3e7, otherParty=Mock Company 1, ourSessionId=6645358286741261650, otherPartySessionId=null, waitingForResponse=true), message=SessionInit(initiatorSessionId=6645358286741261650, initiatorParty=Mock Company 3, flowName=net.corda.flows.NotaryFlow$Client, firstPayload=SignRequest(tx=SignedTransaction(txBits=[120101006A6176612E7574696C2E41727261794C6973F40001000001000101016E65742E636F7264612E636F72652E636F6E7472616374732E5472616E73616374696F6E53746174E50102636F6D2E6469676974616C61737365742E706C6174666F726D2E47736C547853746174E50103636F6D2E6469676974616C61737365742E706C6174666F726D2E47736C436F6E74726163F401046E65742E636F7264612E636F72652E63727970746F2E53656375726548617368245348413235B62190FC9EA356985D4E82B4C7332537CC032B5E244C6C4CCC4DF8002C08AC9496D201056E65742E636F7264612E636F72652E63727970746F2E436F6D706F736974654B6579244C6561E60E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE07401066A6176612E7574696C2E436F6C6C656374696F6E732453696E676C65746F6E4C6973F401050E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE0740101636F7264612E6E6F746172792E73696D706C657C4D6F636B20436F6D70616E7920B101050E20CCFFF852FC48EFCC4508341574BB62D55BDE5B7538025BFB0E6F08DAE1B8D30701000101076E65742E636F7264612E636F72652E636F6E7472616374732E436F6D6D616EE40C0101086E65742E636F7264612E636F72652E63727970746F2E436F6D706F736974654B65F901050E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE0740109636F6D2E6469676974616C61737365742E706C6174666F726D2E47736C47656E65736973436F6D6D616EE414636F7264612E6E6F746172792E73696D706C657C4D6F636B20436F6D70616E7920B101050E20CCFFF852FC48EFCC4508341574BB62D55BDE5B7538025BFB0E6F08DAE1B8D307010601050E204310B62AC8A9DDA637E02834A6820FD596D026E12E925D92910324137DBAE074010A6E65742E636F7264612E636F72652E636F6E7472616374732E5472616E73616374696F6E547970652447656E657261EC00], sigs=[[86595817C69ABFB5070F9C4FAA3B52424F08148AE39A37A8BB08837C4AE6E4870029BFF208CDC08C09A002FCA9910E6A11F6E90A33874DC965B59052161D400C], [4083C5BE0EC4BD7B16EAED7022F74669A335EECBA5AED9636B02CCB3203AE5E08A2FB59146D82D77A4722AF5669A3DC78576F2CFB34B234664944C9205D20407]], id=6CA0AA5854A10E8F341762C62290D456CD219F0D6BB9A6F1A545F76869A7C80A), callerIdentity=Mock Company 3)), receiveType=interface net.corda.node.services.statemachine.StateMachineManager$SessionInitResponse)
    	at net.corda.node.services.statemachine.FlowStateMachineImpl.suspendAndExpectReceive(FlowStateMachineImpl.kt:225) ~[node-0.7-SNAPSHOT.jar:?]
    	at net.corda.node.services.statemachine.FlowStateMachineImpl.access$suspendAndExpectReceive(FlowStateMachineImpl.kt:31) ~[node-0.7-SNAPSHOT.jar:?]
    	at net.corda.node.services.statemachine.FlowStateMachineImpl.startNewSession(FlowStateMachineImpl.kt:297) ~[node-0.7-SNAPSHOT.jar:?]
    	at net.corda.node.services.statemachine.FlowStateMachineImpl.getSession(FlowStateMachineImpl.kt:181) ~[node-0.7-SNAPSHOT.jar:?]
    	at net.corda.node.services.statemachine.FlowStateMachineImpl.sendAndReceive(FlowStateMachineImpl.kt:126) ~[node-0.7-SNAPSHOT.jar:?]
    	at net.corda.core.flows.FlowLogic.sendAndReceive(FlowLogic.kt:59) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.flows.NotaryFlow$Client.call(NotaryFlow.kt:180) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.flows.NotaryFlow$Client.call(NotaryFlow.kt:24) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.core.flows.FlowLogic.subFlow(FlowLogic.kt:90) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.core.flows.FlowLogic.subFlow$default(FlowLogic.kt:84) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.flows.FinalityFlow.call(FinalityFlow.kt:34) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.flows.FinalityFlow.call(FinalityFlow.kt:17) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.core.flows.FlowLogic.subFlow(FlowLogic.kt:90) ~[core-0.7-SNAPSHOT.jar:?]
    	at net.corda.core.flows.FlowLogic.subFlow$default(FlowLogic.kt:84) ~[core-0.7-SNAPSHOT.jar:?]

**Comment by joeldudleyr3** [comment]
Sorry, is there a better way to format that?

**Comment by mikehearn** [comment]
It's formatted fine but isn't a serialisation related problem, so I'm not sure what's happening there. What app is this? Do you have full node logs?

**Comment by joeldudleyr3** [comment]
I'll slack you the app, so you can create the error yourself.

**Comment by adagys** [comment]
@mikehearn it fails on trying to serialize the `MonitoringService`, which is already tagged as `SerializeAsToken`, the error is "All tokens should be registered during context construction.", line 116 SerializationToken.kt
I suspect it's got something to do with how we initialise our mock nodes, can look into it today

**Comment by mikehearn** [comment]
Yeah, that's what I figured was happening - having the Vault object contain an iterable over all unconsumed states was never meant to be a permanent API and this is the second time it's caused trouble in a few days. Let's see if we can kill it off.

**Comment by adagys** [comment]
A quick fix would be removing `asSequence()` from the unconsumed states and not returning a copy of the vault on `notifyAll()`

**Pull request merged**